File: 3686.number-of-stable-subsequences.cpp
#
# @lc app=leetcode id=3686 lang=cpp
#
# [3686] Number of Stable Subsequences
#
# @lc code=start
class Solution {
public:
    int countStableSubsequences(vector<int>& nums) {
        const int MOD = 1e9 + 7;
        int n = nums.size();
        vector<int> dp_odd(n + 1, 0), dp_even(n + 1, 0);
        // Base case initialization for first element
        if (nums[0] % 2 == 0) {
            dp_even[0] = 1;
        } else {
            dp_odd[0] = 1;
        }
        for (int i = 1; i < n; ++i) {
            if (nums[i] % 2 == 0) { // Even number
                dp_even[i] = (dp_even[i - 1] + dp_odd[i - 1]) % MOD;
                // Check for three consecutive evens
                if (i >= 2 && nums[i-1] % 2 == 0 && nums[i-2] % 2 == 0) {
                    dp_even[i] = (dp_even[i] - dp_even[i-2] + MOD) % MOD;
                }
            } else { // Odd number
                dp_odd[i] = (dp_odd[i - 1] + dp_even[i - 1]) % MOD;
                // Check for three consecutive odds
                if (i >= 2 && nums[i-1] % 2 != 0 && nums[i-2] % 2 != 0) {
                    dp_odd[i] = (dp_odd[i] - dp_odd[i-2] + MOD) % MOD;
                }
            }
        }
        long long total_stable_subsequences = accumulate(dp_odd.begin(), dp_odd.end(), static_cast<long long>(0)) \
                                           + accumulate(dp_even.begin(), dp_even.end(), static_cast<long long>(0));
        total_stable_subsequences %= MOD;
        return total_stable_subsequences; // Return total valid subsequences.
    }
};
# @lc code=end

File: 3686.number-of-stable-subsequences.go
#
# @lc app=leetcode id=3686 lang=golang
#
# [3686] Number of Stable Subsequences
#

# @lc code=start
func countStableSubsequences(nums []int) int {
    const MOD = 1e9 + 7
    n := len(nums)
    if n == 0 {
        return 0
    }
    
    dp := make([][2]int, n+1)
    dp[0][0] = 1 // Base case for parity transitions.
    total := int64(0)
    
    for i := range nums {
        parity := nums[i] % 2 // Determine current number's parity.
        
        // Calculate new subsequences ending at position i with current number's parity.
        dp[i+1][parity] = (dp[i][parity^1] + dp[i][parity]) % MOD // Extend both parities without forming invalid triples.
        
        // Correct potential overcount by checking for three consecutive same parity numbers.
        if i >= 2 && nums[i-1]%2 == parity && nums[i-2]%2 == parity {
            dp[i+1][parity] = (dp[i+1][parity] - dp[i-2][parity^1] + MOD) % MOD // Subtract invalid cases correctly.
        }
        
        // Update overall result by adding new valid subsequences ending at current index.
        total = (total + dp[i+1][parity]) % MOD
    }
    return int(total)
}
# @lc code=end